### PR TITLE
feat: handle secret references in FEEL input-mapping evaluation

### LINKED ISSUE #57178: [Secret Resolution] Handle secret references in FEEL input-mapping evaluation
### Description

Handle `camunda.secrets.<name>` references inside FEEL **input-mapping** evaluation so a modeled reference (detected by #56557) survives evaluation as a placeholder instead of failing/nulling — ready for the actual secret value to be substituted at job activation (tracked separately). Two parts:

1. **Deploy-time:** reject `camunda.secrets.<name>` used as a **string literal** in an input mapping.
2. **FEEL evaluation:** evaluate `camunda.secrets.<name>` used as an **expression** as its **string literal** — it produces the text `camunda.secrets.<name>`, not a (null) variable lookup — for input mappings only.

**Constraints that shape the design:**
- Input-mapping FEEL evaluation runs in the engine and its result is **persisted** as local variables → secret **values** must never be persisted; only the reference placeholder is.
- Engine processing must stay **deterministic** (replicated state machine; replay must reproduce state).
- A reference used as an **expression** must be distinguished from the identical text appearing as a **string literal** or a **runtime variable value** (secret-injection prevention).

> Status: **PREFERRED**, pending PM decision.

#### Preferred option — reject literal usage + evaluate the expression reference as a string literal

- **Deploy-time validator:** reject a deployment when `camunda.secrets.<name>` appears as a **string literal** in an input mapping (clear, actionable error). This removes the designer-side literal-vs-expression ambiguity at the source — any `camunda.secrets.<name>` remaining in a valid input mapping is then necessarily an expression.
- **FEEL evaluation:** a secret-aware `EvaluationContext`, installed **only** for input mappings (`BpmnVariableMappingBehavior.applyInputMappings` via `ExpressionProcessor.prependContext`), resolves the path `camunda.secrets.<name>` to the **string** `"camunda.secrets.<name>"`. FEEL leaves real string literals and variable values alone; the expression evaluates to the reference text instead of nulling.
- The actual secret substitution (replacing the placeholder text with the resolved value) happens at **job activation** — tracked in its own issue.

Why preferred:
- **Simple & deterministic.** Evaluating a reference to its own text is a pure transformation — no random, no token, no extra persisted state → trivially replication/replay-safe.
- **Readable placeholder.** Stored variables show `camunda.secrets.<name>` (debuggable), not an opaque token.
- **Ambiguity removed at the source.** Rejecting literal usage means a `camunda.secrets.<name>` in a valid input mapping is always an expression, so the later substitution is unambiguous for the designer-authored case.
- **Right layer.** FEEL natively separates a *path* (code) from a *literal / variable value* (data); only expression paths are affected.

Accepted edge case: a **runtime variable value** equal to `camunda.secrets.<name>` cannot be validated at deploy; at activation it would also be substituted → a **benign double-resolve** of a secret the element already receives (substitution is targeted to the declared paths from #56557; no escalation to other secrets). Documented and tested.

#### Rejected alternatives
- *Nth-occurrence bookkeeping:* the occurrence index is fixed at deploy, but the evaluated string's occurrences depend on runtime variable values (which can inject extra copies and shift indices) → unreliable.
- *Evaluate with the real secret, then persist:* would write secret values into persisted variables → forbidden.
- *Random placeholder generated in processing:* RNG in processing violates the engine's determinism conventions and is fragile on replay; unnecessary here.

#### Possible release-hardening (later, pre-GA) — deterministic token map on deployment

Instead of the readable reference text, mint a **deterministic, opaque token** per reference at **deploy processing** (via `KeyGenerator.nextKey()`, not random) and persist a `token → reference` map with the process definition; the eval context returns the token; job activation replaces tokens using the map.

- Adds robustness against the runtime-variable-value edge — a variable value is far less likely to match an opaque token than the readable reference text.
- Cost: extra persisted state (the token map), deploy-time minting, and less-readable variables.
- Not required for correctness/security given the preferred option already handles the designer case and bounds the runtime edge — so evaluate **before GA** whether the extra robustness is worth the complexity.

### Motivation

Detection (#56557) alone is inert — a modeled `camunda.secrets.<name>` must survive input-mapping evaluation (instead of nulling) and be safe to store as a placeholder, so job activation can substitute the real value. The preferred option achieves this with the least complexity: a deploy-time guard plus a pure, deterministic FEEL evaluation, keeping secret values entirely out of persistence. Recording the reasoning documents why the simpler-looking alternatives fail and what the pre-GA hardening would buy.

### Acceptance criteria

- [ ] Deploy-time validator rejects `camunda.secrets.<name>` used as a string literal in an input mapping, with a clear error.
- [ ] During FEEL evaluation of **input mappings only**, `camunda.secrets.<name>` used as an expression evaluates to the string `"camunda.secrets.<name>"` (no null / no failure).
- [ ] Output mappings, conditions, listeners, etc. are unaffected.
- [ ] Deterministic — no random, no token, no added persisted state; replication/replay-safe.
- [ ] The runtime variable-value collision is documented as an accepted benign edge.
- [ ] Unit tests (reject validation; expression → literal; literal usage rejected; input-mapping-only scope).

### Links

- Parent: #56556
- Depends on: #56557 (secret reference format + parse-time detection)
- Related: job-activation secret substitution (separate issue)

### Breakdown

- Deploy-time validator rejecting literal `camunda.secrets.<name>` in input mappings (design-time and/or runtime validator).
- Secret-aware `EvaluationContext` (input mappings only, via `prependContext`) resolving the reference path to its string literal.
- Tests: reject literal usage; expression evaluates to the reference text; only input mappings affected; determinism.
- (Later / pre-GA hardening) deterministic token map on deployment: mint via `nextKey()`, persist `token → reference`, resolve at activation.

### PR DESCRIPTION
**Pull request by berkaycanbc**
## What

Handle `camunda.secrets.<name>` references in FEEL **input-mapping** evaluation so a modeled
reference survives evaluation as a placeholder instead of nulling — ready for the real secret value
to be substituted at job activation (tracked separately). Two parts:

- **Deploy-time:** reject `camunda.secrets.<name>` used as a **string literal** in an input mapping,
  with a clear error.
- **FEEL evaluation (input mappings only):** `camunda.secrets.<name>` used as an **expression**
  evaluates to the string `"camunda.secrets.<name>"` — no null, no failure.

## Why these decisions

- **Delegation, not shadowing.** The secret-aware context wraps the existing evaluation context and
  intercepts only the `camunda.secrets.*` path. Cluster variables (`camunda.vars.*`), process
  variables and every other lookup are forwarded unchanged, so nothing is shadowed — today or once
  further `camunda.*` namespaces are added.
- **Deterministic.** Resolving a reference to its own text is a pure transformation — no random, no
  token, no added persisted state — so it is replication/replay-safe. Secret *values* are never
  produced or persisted here; only the readable placeholder is.
- **Reject literal usage to remove ambiguity.** Refusing a literal `camunda.secrets.<name>` at
  deploy means any reference remaining in a valid input mapping is necessarily an expression, so the
  later substitution is unambiguous. Detection evaluates the source without a variable context: a
  literal constant-folds to a string containing the reference (including folded concatenations),
  while an expression path resolves to null and is left alone.

## Scope

- Output mappings, conditions, listeners, etc. are unaffected.
- The actual secret-value substitution happens at job activation — separate issue.

## Accepted edge case

A runtime variable *value* equal to `camunda.secrets.<name>` cannot be validated at deploy; at
activation it would also be substituted — a benign double-resolve of a secret the element already
receives. Documented and tested.

## Commits

1. `feat: resolve secret references in input-mapping evaluation`
2. `feat: reject secret reference used as a string literal in input mapping`

Closes #57178

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContext.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  @Override
+  public Either<DirectBuffer, EvaluationContext> getVariable(final String variableName) {
+    if (ROOT.equals(variableName)) {
+      // own the `camunda` segment while keeping the delegate's `camunda` content reachable
+      return Either.right(new CamundaNamespaceContext(delegate.getVariable(ROOT)));
+    }
+    return delegate.getVariable(variableName);
+  }
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContext.java:57]: fixed in 2c1c9955d833c8d6b8a9b8c38c101b71982d5fb2

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContextTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    // then
+    assertThat(result.getString()).isEqualTo("eu-1/camunda.secrets.token");
+  }
+
+  @Test
+  void shouldPassThroughProcessVariables() {
+    // then
+    assertThat(evaluate("=orderId").getString()).isEqualTo("order-123");
+  }
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContextTest.java:99]: fixed in https://github.com/camunda/camunda/commit/2c1c9955d833c8d6b8a9b8c38c101b71982d5fb2

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    // when
+    final var processInstanceKey =
+        ENGINE.processInstance().ofBpmnProcessId("secret-input").create();
+
+    // then
+    final JobRecordValue job =
+        ENGINE.jobs().withType("secret-input-job").activate().getValue().getJobs().getFirst();
+    assertThat(job.getProcessInstanceKey()).isEqualTo(processInstanceKey);
+    assertThat(job.getVariables())
+        .containsEntry("authToken", "camunda.secrets.externalSystemToken");
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java:57]: 💭 When we implement the secret resolution on Jobs this test should still succeed, even if the Job should  receive the Job with resolved secets in there. That's because the engine verifies using exported events and exported events should never get the resolved secret values in it.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    ENGINE.processInstance().ofBpmnProcessId("secret-concat").create();
+
+    // then
+    final JobRecordValue job =
+        ENGINE.jobs().withType("secret-concat-job").activate().getValue().getJobs().getFirst();
+    assertThat(job.getVariables()).containsEntry("authorization", "Bearer camunda.secrets.token");
+  }
+
+  @Test
+  public void shouldNotResolveSecretReferenceInOutputMapping() {
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java:86]: 💭 We definitely need to verify this in an e2e where we don't assert on the exported record. Once we reached that point in implementation.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    final var processInstanceKey =
+        ENGINE.processInstance().ofBpmnProcessId("secret-output").create();
+
+    // when
+    ENGINE.job().ofInstance(processInstanceKey).withType("secret-output-job").complete();
+
+    // then - output mapping did not resolve the reference; the variable is null, not a placeholder
+    final JobRecordValue job =
+        ENGINE.jobs().withType("consumer-job").activate().getValue().getJobs().getFirst();
+    assertThat(job.getVariables()).containsEntry("result", null);
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java:109]: I like the `null` behavior instead of having a reference 👍

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContextTest.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+import io.camunda.zeebe.el.EvaluationResult;
+import io.camunda.zeebe.el.ExpressionLanguage;
+import io.camunda.zeebe.el.ExpressionLanguageFactory;
+import io.camunda.zeebe.el.ResultType;
+import io.camunda.zeebe.engine.processing.bpmn.clock.ZeebeFeelEngineClock;
+import java.time.InstantSource;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+final class SecretReferenceEvaluationContextTest {
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContextTest.java:22]: 🔧 Some other edge-cases we could cover:

- I have a variable named `camunda` and try to reference it
- I have a variable named `camunda.secrets` and try to reference it

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    final Expression expression = expressionLanguage.parseExpression(source);
+    if (!expression.isValid()) {
+      // invalid expressions are reported separately by the expression validator
+      return;
+    }
+
+    // a literal constant-folds without a variable context; an expression path resolves to null
+    final EvaluationResult result =
+        expressionLanguage.evaluateExpression(expression, variable -> Either.left(null));
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java]: ❓ So any variable reference resolves to `null`. Because of this we don' get false-positives when we actually reference a secret which resolves to it's string reference?

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnVariableMappingBehavior.java [hunk 8]
```diff
@@ -44,6 +47,7 @@ public BpmnVariableMappingBehavior(
       final VariableBehavior variableBehavior,
       final EventTriggerBehavior eventTriggerBehavior) {
     this.expressionProcessor = expressionProcessor;
+    inputMappingExpressionProcessor = expressionProcessor.withSecretReferenceContext();
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnVariableMappingBehavior.java:51]: 👍 I like that we only add it to input mappings.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    final var matcher = SECRET_REFERENCE.matcher(constant);
+    if (matcher.find()) {
+      final var reference = matcher.group();
+      validationResultCollector.addError(
+          0,
+          String.format(
+              "Secret reference '%s' must be used as an expression (e.g. '=%s'), not as a string"
+                  + " literal, in input mapping source '%s'.",
+              reference, reference, source));
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java]: 🔧 If a user has multiple String literal secret names we only return the name of the first one. It would be nice if the error message contained all so the user doesn't have to do another round-trip before discovering a second reference.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+import io.camunda.zeebe.el.ExpressionLanguage;
+import io.camunda.zeebe.el.ExpressionLanguageFactory;
+import io.camunda.zeebe.engine.processing.bpmn.clock.ZeebeFeelEngineClock;
+import io.camunda.zeebe.model.bpmn.instance.zeebe.ZeebeInput;
+import java.time.InstantSource;
+import org.camunda.bpm.model.xml.validation.ValidationResultCollector;
+import org.junit.jupiter.api.Test;
+
+final class SecretReferenceLiteralValidatorTest {
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidatorTest.java:26]: 🔧 Can we add some strange characters here?
- '`'
- Some random UTF characters?

With the goal of verifying the regex properly captures all scenarios.

**Review by remcowesterhoud — APPROVED** [review]
@berkaycanbc I have a few comments, please have a look at them. Nothing blocking the merge.

-----

ℹ️ Just to be make it clear. We do break backwards compatibility here for users who already use `camunda.secrets.X` as a variable naming pattern. We've done this in the past and accepted this.

We also break for users referencing string literals `camunda.secrets.X`, by rejecting their deployment.

@berkaycanbc Just to make sure, we've got an explicit go-ahead that this is fine?

-------

💭 We should ensure we have some tests cases where we have a variable named `camunda.secrets.token`, or more complext:

```json
{
  camunda: {
    secets: {
      token: "foo"
    }
  }
}
```

Not really part of this PR but just something I think we should cover in our test to ensure the behavior is what we'd expect. It'll probably work just fine, except you can't really reference it in a FEEL expression as we'd replace the value.

What's also a good edge-case to cover in our tests is if we have a variable names as `"camunda.secret.token" ` (string) to ensure we don't replace this variable name with the resolved secret.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java (hunk 9 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java]: Done in https://github.com/camunda/camunda/commit/ff25f5b879a2687bd5b5b00e4cf073e6c0b05f00. All distinct references are now collected and listed in a single error, so a user doesn't have to redeploy to discover a second one.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidatorTest.java (hunk 10 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidatorTest.java:26]: Added in https://github.com/camunda/camunda/commit/ff25f5b879a2687bd5b5b00e4cf073e6c0b05f00 — the regex is now unicode-aware (`[\p{L}\p{N}_]`), with tests for a unicode name (`camunda.secrets.tökén`) and for digits/underscores. Backtick-quoted names only arise in the FEEL expression form (which is allowed and resolves to the placeholder), not in a rejected string literal, so there's nothing extra to catch there.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContextTest.java (hunk 6 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceEvaluationContextTest.java:22]: Covered in https://github.com/camunda/camunda/commit/9baa0e103e1625d73ab32c3c59f48e8b5c92bbae. Added a test where `camunda.secrets.<name>` is modeled as a real object variable and the variable wins (its value is returned, not the placeholder). Together with the existing tests for a variable named `camunda` (`shouldNotShadowProcessVariableNamedCamunda`, `shouldNotResolveSecretWhenCamundaVariablePresent`) and a variable *value* equal to a reference (`shouldTreatVariableValueEqualToReferenceAsPlainString`), all three edge cases are now locked in.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java (hunk 7 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/SecretReferenceLiteralValidator.java]: Correct, no false positives. The validator evaluates the source with an empty context, so a real expression like `=camunda.secrets.token` resolves to `null` (not a constant string/object/array) and is never flagged; only a genuine literal constant-folds to a value that contains the reference text. Covered by `shouldAllowSecretReferenceUsedAsAnExpression`.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java (hunk 3 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java:57]: This test asserts on exported records, which by design never carry resolved secret values, so it stays valid once job-level secret resolution lands. 👍

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java (hunk 4 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/expression/SecretReferenceInputMappingTest.java:86]: Makes sense. Let's add an activation-level e2e that inspects the delivered job (not the exported record) once the job-activation substitution is implemented, tracked in its own issue.

**Comment by berkaycanbc** [comment]
> @berkaycanbc I have a few comments, please have a look at them. Nothing blocking the merge.
> 
> ℹ️ Just to be make it clear. We do break backwards compatibility here for users who already use `camunda.secrets.X` as a variable naming pattern. We've done this in the past and accepted this.
> 
> We also break for users referencing string literals `camunda.secrets.X`, by rejecting their deployment.
> 
> @berkaycanbc Just to make sure, we've got an explicit go-ahead that this is fine?

@remcowesterhoud Validated this with @StevePascoe and yes.
